doc/tags_param.md

```diff
@@ -0,0 +1,205 @@
+gkernel.hasInt("log_width")     # returns false
+gkernel.hasFloat("log_width")   # returns true
+gkernel.hasFloat("foo")     # returns false
+```
+
+The semantics remain the same with minute syntactic differences. Important things to note:
+- `sets()` can be used to set a parameter using its name or a `Tag<Type>` (like `TagFloat`) object.
+- `gets()` can be used to query a parameter's value using a `Tag<Type>` (like `TagFloat`) object **but** `gets<Type>()` (like `getsFloat()`) is used to query the value by the parameter's name.
+- `has()` can be used to check the availability of a parameter by its name. `has<Type>()` (like `hasFloat()`) is used to check the availability of a parameter by it's name and type.
+
+These nomenclature of functions `sets`/`gets`/`has` is same in all the high level languages supported by Shogun i.e. Python, Ruby, R, Lua, Java, C# and Octave.
+
+*A pro tip:* Use `gets<Tab>`/`has<Tab>`/`Tag<Tab>` auto-completion to list all the types available.
+
+### For developers
+
+All this is accomplished by having a `std::unordered_map<BaseTag, Any>` (or `std::map<BaseTag, Any>` depending on the availability of C++11) in `SGObject.cpp`, where the map stores the values of parameters available in a Shogun class derived from `CSGObject`. From here on, I'll refer to the `SGObject`'s `std::unordered_map`/`std::map` as `map`.
+
+#### BaseTag
+
+`BaseTag` is the parent class of `Tag`. For a particular parameter, `BaseTag` stores the name, while `Tag` (being a template class) stores the type information. Name and type information are stored separately because `BaseTag` also stores a hash generated from the parameter's name. `BaseTag` is used as the key in the `map`, this results in fast look-ups.
+
+See examples of `BaseTag` in `SGObject.h`.
+
+#### Any
+
+`Any` allows to store objects of arbitrary types in a type agnostic way. This makes it possible to store a variety of types in the `map`.
+
+```cpp
+int32_t integer = 5;                // integer
+GaussianKernel gkernel(5);          // gaussian kernel object
+auto any_integer = Any(integer);    // any object for integer
+auto any_kernel = Any(gkernel);     // any object of gaussian kernel
+```
+
+`Tag`, `BaseTag` and `Any` can be found in `src/shogun/lib`.
+
+#### register_param() / register_member()
+
+While defining a new Shogun class, `register_param()` or `register_member()` should be used in the constructor to register parameters and member variables of the class. This would allow the parameters or class member variables values to be queried or modified using `gets()` / `sets()`. These functions are required to prevent Shogun users to modify and query arbitrary parameters in an object but only those set of parameters which have been registered by using these two functions. The new Shogun class should also inherit `CSGObject` as these two functions are `protected` in `CSGObject`. Let's look at an example to make this more concrete.
+
+```cpp
```

Comment:
@lisitsyn I wrote the above paragraph and realized then we are now going to use `SG_ADD` which only has `register_member()`. How will `register_param()` be used? Do we need it anymore? Will the new plugin classes also use `SG_ADD`?

@OXPHOS This section is incomplete. Will complete it soon.